### ISSUE TITLE
libtinfo required when running in WSL

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1305]
https://github.com/simonmichael/hledger/pull/1075#issuecomment-663845226

**Comment by adept** [comment]
I think that comment in #1075 has nothing to do with Docker. 

It looks like an attempt to run a pre-built binary in WSL, when `libtinfo` was not installed, and whoever hit that issue just searched for `hledger tinfo`, found old closed #1075 that *also* mentioned `libtinfo` (even though it was in a completely different context) and added their comment there.

**Comment by simonmichael** [comment]
@artkpv, when you get a chance please explain your problem, otherwise we'll close this in a bit.

**Comment by artkpv** [comment]
@simonmichael, Yes, @adept is right in that I tried to run hledger binaries in WSL 2 with Ubuntu. It reported that no libtinfo.so (see message in the comment). Probably Hledger has libtinfo5 dependency to install?

**Comment by adept** [comment]
@artkpv : where did you get the binary from? Was it package manager (`apt-get install`) or pre-built binaries from hledger.org?

If it is the latter, then there is no way to trigger or enforce installation of `libtinfo5` when you start using them

**Comment by adept** [comment]
I re-read @artkpv's comment - version 1.18.99 clearly indicates that it was not from the package manager. As such I don't think there is anything else to do here - we can't force OS to install missing libraries

**Comment by artkpv** [comment]
@adept  It was pre-built. I downloaded it from hledger.org.

**Comment by adept** [comment]
@simonmichael i wonder if it worthwhile mentioning on the download page that with prebuilt binaries you need to take care of all the required libs yourself?

**Comment by simonmichael** [comment]
@artkpv so, the hledger-ubuntu.zip at https://hledger.org/download.html, right ? 

That links to https://github.com/simonmichael/hledger/releases/tag/1.18.1 which says "Ubuntu binaries are dynamically linked x86_64 binaries built on ubuntu 20.04, depending on these libs: linux-vdso.so.1, libm.so.6, libtinfo.so.5, librt.so.1, libutil.so.1, libdl.so.2, libpthread.so.0, libgmp.so.10, libc.so.6." 

We could improve the wording there. But I don't want this page to become a whole 'nother user install guide (which has to be recreated for every release). Any suggestions welcome.

**Comment by simonmichael** [comment]
I've added a link from there to https://hledger.org/download.html#ensure-c-libraries-are-installed. Maybe that helps.

**Comment by simonmichael** [comment]
I'll improve the wording at https://hledger.org/download.html#ensure-c-libraries-are-installed if folks can help me understand the situation better.  I think 

- libtinfo-dev (eg) provides some terminfo library 
- that's required to build hledger CLI
- and also to run the hledger CLI binary from the CI-generated hledger-ubuntu.zip

Why do we have to install the -dev library, and not just "libtinfo" ?

**Comment by adept** [comment]
I claim that ther eis no need to install -dev, unless you plan to compile hledger

**Comment by simonmichael** [comment]
Doesn’t @artkpv’s report show otherwise, if I’m interpreting it correctly ?

**Comment by adept** [comment]
Here is how i read it:
Original docker ticker was for libtinfo-dev required to fix the build in docker and installation of lbtinfo-dev is mentioned as a fix. @artkpv says "I tried installing libtinfo-dev and it did not help".

How do you read it?

**Comment by simonmichael** [comment]
I'm reading the "I tried to run hledger binaries in WSL 2 with Ubuntu. It reported that no libtinfo.so" comment.

**Comment by adept** [comment]
> I'll improve the wording at https://hledger.org/download.html#ensure-c-libraries-are-installed if folks can help me understand the situation better. I think
> 
> * libtinfo-dev (eg) provides some terminfo library
> * that's required to build hledger CLI
> * and also to run the hledger CLI binary from the CI-generated hledger-ubuntu.zip
> 
> Why do we have to install the -dev library, and not just "libtinfo" ?

https://hledger.org/download.html#ensure-c-libraries-are-installed is in the section called `Building from sources`, so for building the advice to install libtinfo-dev is correct (it provides .a, but not .so).

For running the binary, it is enough to have `libtinfo5` installed:

```
$ ldd ~/.local/bin/hledger
	linux-vdso.so.1 (0x00007ffe93b04000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8e10eba000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f8e10c90000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f8e10a88000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f8e10885000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8e10681000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8e10462000)
	libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f8e101e1000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8e0fdf0000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8e11258000)

$ dlocate /lib/x86_64-linux-gnu/libtinfo.so.5
libtinfo5:amd64: /lib/x86_64-linux-gnu/libtinfo.so.5.9
libtinfo5:amd64: /lib/x86_64-linux-gnu/libtinfo.so.5
``` 

Here is whats in dev:
```
$ dlocate libtinfo-dev
libtinfo-dev:amd64: /.
libtinfo-dev:amd64: /usr
libtinfo-dev:amd64: /usr/lib
libtinfo-dev:amd64: /usr/lib/valgrind
libtinfo-dev:amd64: /usr/lib/valgrind/ncurses.supp
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtermcap.so
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtic.a
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtinfo.a
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/pkgconfig
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/pkgconfig/tic.pc
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/pkgconfig/tinfo.pc
libtinfo-dev:amd64: /usr/share
libtinfo-dev:amd64: /usr/share/doc
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtermcap.a
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtic.so
libtinfo-dev:amd64: /usr/lib/x86_64-linux-gnu/libtinfo.so
libtinfo-dev:amd64: /usr/share/doc/libtinfo-dev
```

.so listed there is a symlink, not a file

**Comment by simonmichael** [comment]
That helps, thanks @adept. So our advice could be: if running the hledger-ubuntu CI binary, apt install libtinfo5; if building hledger, apt install libtinfo5-dev.

**Comment by adept** [comment]
Yep

**Comment by artkpv** [comment]
Thanks for this great efforts to build hledger. 

> @artkpv so, the hledger-ubuntu.zip at https://hledger.org/download.html, right ?

Yes, that's right

Steps were:
0. Windows 10 with WSL 2 and Ubuntu on it.
1. Download from binaries from https://hledger.org/download.html. Didn't use apt as version is lower.
2. Tried to run it. It reported error.
3. Read on the Github release page. Followed https://hledger.org/download.html#ensure-c-libraries-are-installed. 
4. Run `sudo apt install libtinfo-dev`
5. Tried hledger. Didn't work:

> hledger: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory

6. Search hledger Github. Found this issue https://github.com/simonmichael/hledger/pull/1075
7. Installed `sudo apt install libtinfo5`

I don't know the actual differences between libtinfo5 and libtinfo-dev. No sense to me. Hope it helps

**Comment by simonmichael** [comment]
Yes that does help, thanks @artkpv.

**Comment by simonmichael** [comment]
I have promoted the https://hledger.org/download.html#c-libraries section and clarified that it's required both for running prebuilt binaries and building from source. 

The list of required packages is probably incomplete, but it's what I know for now. To keep things simple, I've just listed the packages required for building (eg libtinfo-dev), which also pulls in packages needed for running (libtinfo5).

I think we can close this now.
